runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryStateInternals.java

```diff
@@ -463,6 +475,114 @@ public InMemoryBag<T> copy() {
     }
   }
 
+  /** An {@link InMemoryState} implementation of {@link MultimapState}. */
+  public static final class InMemoryMultimap<K, V>
+      implements MultimapState<K, V>, InMemoryState<InMemoryMultimap<K, V>> {
+    private final Coder<K> keyCoder;
+    private final Coder<V> valueCoder;
+    private Multimap<K, V> contents = ArrayListMultimap.create();
```

Comment:
Using a map that stores keys only works if the key coder is consistent with equals otherwise the keys that you need to use are the `Coder#structuralValue`. This is a bug in the InMemoryMap implementation as well.

You can test this by using `byte[]` keys that have the same contents but are different instances.

Reply:
Ah I didn't think of this before. Then this means beam Map/Multimap will have a different semantic from the java Map: 

* in beam Map/Multimap: 2 keys are considered same key if their encoded byte array is the same.
* in java Map: 2 keys are considered same key if `key1.equals(key2)`

Reply:
I have updated the code to use `Coder#structuralValue` and added a test case for that.